### ISSUE TITLE
[upstart] docker.conf is already provided by package, conflicting

### ISSUE THREAD

**Issue by scalp42** [chef-cookbooks/docker#647]
Given that we're overriding the upstart conf and you upgrade docker-engine through apt, you'll run into the dpkg issue around the already existing upstart.conf not matching the one inside the package:

```
[2016-02-16T22:29:47+00:00] INFO: Sending resource update report (run-id: 661aed7e-ca8e-4b1f-a7e0-bebcbf411d15)
[2016-02-16T22:29:47+00:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
[2016-02-16T22:29:47+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2016-02-16T22:29:47+00:00] ERROR: docker_service[default] (gz-docker::default line 38) had an error: Mixlib::ShellOut::ShellCommandFailed: docker_installation_package[default] (/var/chef/cache/cookbooks/docker/libraries/docker_service.rb line 61) had an error: Mixlib::ShellOut::ShellCommandFailed: apt_package[docker-engine] (/var/chef/cache/cookbooks/docker/libraries/docker_installation_package.rb line 16) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '100'
---- Begin output of apt-get -q -y install docker-engine=1.10.1-0~trusty ----
STDOUT: Reading package lists...
Building dependency tree...
Reading state information...
The following extra packages will be installed:
  libsystemd-journal0
Recommended packages:
  yubico-piv-tool
The following NEW packages will be installed:
  libsystemd-journal0
The following packages will be upgraded:
  docker-engine
1 upgraded, 1 newly installed, 0 to remove and 61 not upgraded.
Need to get 8746 kB of archives.
After this operation, 5346 kB of additional disk space will be used.
Get:1 https://apt.dockerproject.org/repo/ ubuntu-trusty/main docker-engine amd64 1.10.1-0~trusty [8696 kB]
Get:2 http://us-west-2.ec2.archive.ubuntu.com/ubuntu/ trusty-updates/main libsystemd-journal0 amd64 204-5ubuntu20.18 [49.9 kB]
Fetched 8746 kB in 2s (3231 kB/s)
Selecting previously unselected package libsystemd-journal0:amd64.
(Reading database ... 181599 files and directories currently installed.)
Preparing to unpack .../libsystemd-journal0_204-5ubuntu20.18_amd64.deb ...
Unpacking libsystemd-journal0:amd64 (204-5ubuntu20.18) ...
Preparing to unpack .../docker-engine_1.10.1-0~trusty_amd64.deb ...
docker stop/waiting
Unpacking docker-engine (1.10.1-0~trusty) over (1.9.1-0~trusty) ...
Processing triggers for man-db (2.6.7.1-1ubuntu1) ...
Processing triggers for ureadahead (0.100.0-16) ...
Setting up libsystemd-journal0:amd64 (204-5ubuntu20.18) ...
Setting up docker-engine (1.10.1-0~trusty) ...
Processing triggers for libc-bin (2.19-0ubuntu6.6) ...
STDERR: W: Duplicate sources.list entry https://apt.dockerproject.org/repo/ ubuntu-trusty/main amd64 Packages (/var/lib/apt/lists/apt.dockerproject.org_repo_dists_ubuntu-trusty_main_binary-amd64_Packages)

Configuration file '/etc/init/docker.conf'
 ==> Modified (by you or by a script) since installation.
 ==> Package distributor has shipped an updated version.
   What would you like to do about it ?  Your options are:
    Y or I  : install the package maintainer's version
    N or O  : keep your currently-installed version
      D     : show the differences between the versions
      Z     : start a shell to examine the situation
 The default action is to keep your current version.
*** docker.conf (Y/I/N/O/D/Z) [default=N] ? dpkg: error processing package docker-engine (--configure):
 EOF on stdin at conffile prompt
Errors were encountered while processing:
 docker-engine
W: Duplicate sources.list entry https://apt.dockerproject.org/repo/ ubuntu-trusty/main amd64 Packages (/var/lib/apt/lists/apt.dockerproject.org_repo_dists_ubuntu-trusty_main_binary-amd64_Packages)
W: You may want to run apt-get update to correct these problems
E: Sub-process /usr/bin/dpkg returned an error code (1)
---- End output of apt-get -q -y install docker-engine=1.10.1-0~trusty ----
Ran apt-get -q -y install docker-engine=1.10.1-0~trusty returned 100
[2016-02-16T22:29:47+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

I'll open a PR shortly.

**Comment by someara** [comment]
Are you on the latest version?
I completely revamped the upstart bits to avoid this last week

**Comment by someara** [comment]
Script template from packages:
https://github.com/chef-cookbooks/docker/blob/master/templates/default/upstart/docker.conf.erb

Extra "don't clobber the package" gaurds -
https://github.com/chef-cookbooks/docker/blob/master/libraries/docker_service_manager_upstart.rb#L19

**Comment by scalp42** [comment]
Using: 

```
root@mesos-master-i-0b2ca9d1# cat /var/chef/cache/cookbooks/docker/metadata.json | jq .version
"2.4.24"
```

**Comment by scalp42** [comment]
Weird but I'll fix it by hand. Hitting this when upgrading from 1.9.x to 1.10.1.

**Comment by scalp42** [comment]
[Another one](http://www.thecoli.com/media/dj-khaled-another-one.6093/full):

```
root@mesos-master-i-635669ba # apt-get install -f
Reading package lists... Done
Building dependency tree
Reading state information... Done
0 upgraded, 0 newly installed, 0 to remove and 61 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Setting up docker-engine (1.10.1-0~trusty) ...

Configuration file '/etc/init/docker.conf'
 ==> Modified (by you or by a script) since installation.
 ==> Package distributor has shipped an updated version.
   What would you like to do about it ?  Your options are:
    Y or I  : install the package maintainer's version
    N or O  : keep your currently-installed version
      D     : show the differences between the versions
      Z     : start a shell to examine the situation
 The default action is to keep your current version.
```

I think we should allow passing dpkg options, it doesn't hurt anything. What do you think?

**Comment by someara** [comment]
The cookbook only recently changed the script sources...

You probably have an old version of the upstart script and the cookbook
isn't clobbering it with the new one because it's not supposed to anymore

-s

**Comment by scalp42** [comment]
Right, but if it's not clobbering it with the new one, then on next package upgrade, dpkg will complain, which is the issue.

**Comment by someara** [comment]
See if commenting out the not_if on the template resource makes things
smoother

**Comment by horkhe** [comment]
The thing is that the file had been clobbered by the older version of this cookbook when 1.9.1 was provisioned in the first place. So when we try to upgrade to 1.10.1 apt-get complains regardless of the cookbook version we are using.

If I understood you correctly the latest version does not modify those files. So it is safe to pick option `Y`, or rather run `apt-get` with `-o Dpkg::Options::=--force-confdef` when you install the package.

**Comment by scalp42** [comment]
Right @horkhe I think just adding the possibility to pass options would fix the issue. I'm going to install docker-engine by hand to get the upstart conf from the package in the meantime and then re-run docker cookbook. Not ideal, but right now it's failing.

**Comment by horkhe** [comment]
@scalp42 yea, installing docker by hand is what I ended up doing.

I am not a big fan of allowing to pass arbitrary options though. It looks like an abstraction leak to me. I would prefer introducing `overwirte_modified_files` boolean parameter to `docker_installation_package` LWRP. But that is just me, and it is up to @someara to decide. Afterall either solution would do the trick.

**Comment by scalp42** [comment]
@horkhe I'd argue on that because `options` is just leveraging the underlying package resource in Chef directly, but either way :sake:

**Comment by someara** [comment]
I've merged the package_options PR and released it in 2.5.1

**Comment by scalp42** [comment]
Thanks a lot @someara !
